vivarium/core/registry.py

```diff
@@ -335,117 +353,129 @@ def divide_null(state):
-        this function.
-
+        """Serialize data using correct codec.
+        
+        This is typically only called in the case that individual stores
+        are assigned custom serializers. For maximum efficiency, serialization
+        should be left to ``pymongo`` instead of calling this function.
+        
+        Do not override this function without good reason.
+        
         Args:
-            data: Data to serialize.
-
+            data: Data to serialize
+            
         Returns:
-            The serialized data.
-        """
-        string_serialization = self.serialize_to_string(data)
-        if not isinstance(string_serialization, str):
-            raise ValueError(
-                f'{self}.serialize_to_string() returned invalid '
-                f'serialization: {string_serialization}')
-
-        return f'!{self.name}[{string_serialization}]'
-
-    def can_serialize(self, data):
-        """Check whether the serializer can serialize some data.
-
-        Every subclass must implement this method.
+            Serialized data
         """
-        raise NotImplementedError(
-            f'{self} has not implemented can_serialize().')
+        for codec in self.codecs:
+            if isinstance(data, codec.python_type):
+                return codec.transform_python(data)
 
     def deserialize_from_string(self, data):
         """Deserialize data from a string.
 
-        This function is the inverse of
-        :py:meth:`vivarium.core.registry.Serializer.serialize_to_string`.
-        It need only be overridden in a subclass if that subclass does
-        not override
-        :py:meth:`vivarium.core.registry.Serializer.deserialize`.
+        This function should only be overwritten if a codec does not form
+        a one-to-one mapping between BSON and Python types and instead
+        serializes data to strings of the form::
+
+            f'!{self.name or self.__class__.__name__}[serialized_data]'
+
+        It should turn ``serialized_data`` into the correct Python type.
+        
+        Args:
+            data: Serialized data without the 
+                ``!{self.name or self.__class__.__name__}``
+                prefix or the brackets
+                
+        Returns:
+            The deserialized data
         """
         raise NotImplementedError(
             f'{self} has not implemented deserialize_from_string().')
 
     def deserialize(self, data):
         """Deserialize some data.
 
-        Subclasses relying on
-        :py:meth:`vivarium.core.registry.Serializer.serialize_to_string`
-        should not override this function and instead override
+        This function should typically only be overwritten if all codecs in
+        the subclass form one-to-one mappings between BSON and Python types.
+        Otherwise, leave this function alone and only override
         :py:meth:`vivarium.core.registry.Serializer.deserialize_from_string`.
```

Comment:
In this case, the function should just `pass`, right? It seems strange to ask users to override a function just to make it `pass`. Can we have an attribute that specifies whether this serializer is handled by Pymongo or not and then have this function check that attribute?

Reply:
Cleaned up API significantly in 5394975.